### PR TITLE
chore: Inject the events.Recorder into the context.Context

### LINKED ISSUE #304: Inject the `events.Recorder` into the `context.Context`
### Tell us about your request

It would be nice if we didn't have to pass the event recorder through every time that we wanted to use it in any of the standard contexts. Instead, we could do something like `recorder.FromContext(ctx).Publish()`

### Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?

N/A

### Are you currently working around this issue?

Passing the event recorder around through functions anytime that it's needed

### Additional Context

_No response_

### Attachments

_No response_

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### PR DESCRIPTION
**Pull request by sadath-12**
Fixes #304  

**Description**

**How was this change tested?**

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by sadath-12** [comment]
Not completed !! for the review

**Comment by tzneal** [comment]
I don't see that this is performance related. The docs for context state:

> Package context defines the Context type, which carries deadlines, cancellation signals, and other request-scoped values across API boundaries and between processes. 

The event recorder isn't request-scoped, so I don't think it really belongs in the context.

**Comment by sadath-12** [comment]
Any suggestions on what approach to be taken to close issue #304

File: pkg/operator/operator.go [hunk 1]
```diff
@@ -183,10 +183,11 @@ func NewOperator() (context.Context, *Operator) {
 	lo.Must0(mgr.AddHealthzCheck("healthz", healthz.Ping))
 	lo.Must0(mgr.AddReadyzCheck("readyz", healthz.Ping))
 
+	ctx = events.ToContext(ctx,events.NewRecorder(mgr.GetEventRecorderFor(appName)))
```
Comment (jonathan-innis) [inline pkg/operator/operator.go:186]: Nice 🎉

File: pkg/controllers/provisioning/suite_test.go [hunk 2]
```diff
@@ -71,7 +69,7 @@ var _ = BeforeSuite(func() {
 	fakeClock = clock.NewFakeClock(time.Now())
 	cluster = state.NewCluster(fakeClock, env.Client, cloudProvider)
 	nodeController = informer.NewNodeController(env.Client, cluster)
-	prov = provisioning.NewProvisioner(env.Client, corev1.NewForConfigOrDie(env.Config), events.NewRecorder(&record.FakeRecorder{}), cloudProvider, cluster)
+	prov = provisioning.NewProvisioner(env.Client, corev1.NewForConfigOrDie(env.Config), cloudProvider, cluster)
```
Comment (jonathan-innis) [inline pkg/controllers/provisioning/suite_test.go:72]: I think we may need to inject a recorder into the context generally for all of our test environments. If we don't pass it through, I think these tests will panic when we try to fire errors

File: pkg/controllers/provisioning/controller.go [hunk 3]
```diff
@@ -53,7 +52,7 @@ func (c *Controller) Name() string {
 }
 
 // Reconcile the resource
-func (c *Controller) Reconcile(_ context.Context, p *v1.Pod) (reconcile.Result, error) {
+func (c *Controller) Reconcile(ctx context.Context, p *v1.Pod) (reconcile.Result, error) {
```
Comment (jonathan-innis) [inline pkg/controllers/provisioning/controller.go:55]: Why does this need to be changed?

File: pkg/controllers/provisioning/controller.go [hunk 4]
```diff
@@ -40,11 +40,10 @@ type Controller struct {
 }
 
 // NewController constructs a controller instance
-func NewController(kubeClient client.Client, provisioner *Provisioner, recorder events.Recorder) corecontroller.Controller {
+func NewController(kubeClient client.Client, provisioner *Provisioner) corecontroller.Controller {
 	return corecontroller.Typed[*v1.Pod](kubeClient, &Controller{
 		kubeClient:  kubeClient,
 		provisioner: provisioner,
-		recorder:    recorder,
```
Comment (jonathan-innis) [inline pkg/controllers/provisioning/controller.go:47]: I'm imagining that this recorder can just be dropped outright without any other changes to this file

File: pkg/cloudprovider/types.go [hunk 5]
```diff
@@ -137,7 +137,8 @@ type Offerings []Offering
 // passed zone and capacity type
 func (ofs Offerings) Get(ct, zone string) (Offering, bool) {
 	return lo.Find(ofs, func(of Offering) bool {
-		return of.CapacityType == ct && of.Zone == zone
+		return of.CapacityType == ct && of.Zone ==
+		 zone
```
Comment (jonathan-innis) [inline pkg/cloudprovider/types.go:141]: Why did we need to add a newline here?

**Comment by sadath-12** [comment]
here there is problem , we have to pass the recorder to the baseContext but recorder depends on mgr which comes after baseContext

File: pkg/operator/operator.go (hunk 1 above)
Comment (sadath-12) [reply pkg/operator/operator.go:186]: here there is problem , we have to pass the recorder to the baseContext but recorder depends on mgr which comes after baseContext

File: pkg/controllers/provisioning/suite_test.go (hunk 2 above)
Comment (sadath-12) [reply pkg/controllers/provisioning/suite_test.go:72]: ya thats what thinkered me

File: pkg/controllers/provisioning/controller.go (hunk 3 above)
Comment (sadath-12) [reply pkg/controllers/provisioning/controller.go:55]: By miss

**Comment by jonathan-innis** [comment]
> also what is the hack that we can do for

I'm not sure that there is a good hack here. Considering what you are saying about the baseContext, I'm not convinced that we should do the injection. I'm more convinced that we should just leave the implementation as-is. Thoughts?

**Comment by sadath-12** [comment]
> > also what is the hack that we can do for
> 
> I'm not sure that there is a good hack here. Considering what you are saying about the baseContext, I'm not convinced that we should do the injection. I'm more convinced that we should just leave the implementation as-is. Thoughts?

Ya it was my concern that avoiding that injection might break something . Ya ill leave that decision on you

**Comment by jonathan-innis** [comment]
> Ya ill leave that decision on you

I think given the bit that you've discovered about the ordering of when the recording gets instantiated. I'm inclined to close this PR and close the issue as a `wont-fix`. I'm going to close both for now. Please feel free to re-open if you think one of them was closed by mistake.

**Pull request closed without merging**